apps/emqx_utils/src/emqx_utils_fmt.erl

```diff
@@ -0,0 +1,278 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_utils_fmt).
+
+-compile(export_all).
+-compile(nowarn_export_all).
+
+-export([
+    table/2,
+    table/3
+]).
+
+%%
+
+-type header_cell() :: unicode:chardata() | {unicode:chardata(), column_props()}.
+-type table_cell() :: table_cell_simple() | table_sub_structure().
+-type table_cell_simple() :: unicode:chardata() | atom() | number().
+-type table_sub_structure() ::
+    {subrows, [[table_cell()]]}
+    | {subcolumns, [[table_cell_simple()]]}
+    | {group, [[table_cell()]]}.
+
+-type table_props() :: #{_ => _}.
+-type column_props() :: #{align => left | right}.
+-type cell_props() :: #{padc => char(), delimiter => string(), _ => _}.
+
+-type formatted_cell() :: unicode:chardata() | {unicode:chardata(), cell_props()}.
+
+-define(table_header_props, #{padc => $-, delimiter => "."}).
+-define(table_headcut_props, #{padc => $-}).
+-define(group_header_props, #{padc => $-}).
+-define(table_footer_props, #{padc => $-, delimiter => "`"}).
+
+%%
+
+%% @doc Format a set of rows accompanied by header into a printable, ASCII-delimited
```

Comment:
Is terminal width taken into consideration? 
Wonder if JSON format is good enough, some other tools can help to format JSON object array into tables .